File: Ahorn/entities/forceDashCrystal.jl
module ForceDashCrystal

using ..Ahorn, Maple

@mapdef Entity "vitellary/forcedashcrystal" Crystal(x::Integer, y::Integer, oneUse::Bool=false,
    direction::String="Right", respawnTime::Number=2.5, needDash::Bool=false)

const directions = ["Right", "Downright", "Down", "Downleft", "Left", "Upleft", "Up", "Upright", "None"]

const directionIsOrtho = Dict{String, Bool}(
	"Right" => true,
	"Downright" => false,
	"Down" => true,
	"Downleft" => false,
	"Left" => true,
	"Upleft" => false,
	"Up" => true,
	"Upright" => false,
    "None" => false
)

const rotations = Dict{String, Number}(
	"Right" => pi * 0.5,
	"Downright" => pi * 0.5,
	"Down" => pi,
	"Downleft" => pi,
	"Left" => -pi * 0.5,
	"Upleft" => -pi * 0.5,
	"Up" => 0,
	"Upright" => 0
)

const placements = Ahorn.PlacementDict(
    "Force Dash Crystal ($(dir)) (Crystalline)" => Ahorn.EntityPlacement(
        Crystal,
		"point",
		Dict{String, Any}(
            "direction" => dir
        )
    ) for dir in directions
)

Ahorn.editingOptions(entity::Crystal) = Dict{String, Any}(
    "direction" => directions
)

function Ahorn.selection(entity::Crystal)
    x, y = Ahorn.position(entity)

    return Ahorn.Rectangle(x-8, y-8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Crystal)
	dir = get(entity.data, "direction", "Right")
    needdash = get(entity.data, "needDash", false)
    crystalType = needdash ? "needdash" : "dashless"
    sprite = "objects/crystals/forcedash/" * crystalType * "/" * (directionIsOrtho[dir] ? "ortho" : "diag") * "/idle00.png"
	
    if dir == "None"
        Ahorn.Cairo.save(ctx)
        
        Ahorn.drawSprite(ctx, sprite, 2, -2, jx=0.5, jy=0.5)
        
        Ahorn.Cairo.rotate(ctx, pi * 0.5)
        Ahorn.drawSprite(ctx, sprite, 2, -2, jx=0.5, jy=0.5)
        
        Ahorn.Cairo.rotate(ctx, pi * 0.5)
        Ahorn.drawSprite(ctx, sprite, 2, -2, jx=0.5, jy=0.5)
        
        Ahorn.Cairo.rotate(ctx, pi * 0.5)
        Ahorn.drawSprite(ctx, sprite, 2, -2, jx=0.5, jy=0.5)
            
        Ahorn.Cairo.restore(ctx)
    else
        Ahorn.Cairo.save(ctx)
        Ahorn.Cairo.rotate(ctx, rotations[dir])
        if directionIsOrtho[dir]
            Ahorn.drawSprite(ctx, sprite, 0, -2, jx=0.5, jy=0.5)
            Ahorn.drawSprite(ctx, sprite, 4, 2, jx=0.5, jy=0.5)
            Ahorn.drawSprite(ctx, sprite, -4, 2, jx=0.5, jy=0.5)
        else
            Ahorn.drawSprite(ctx, sprite, 2, -2, jx=0.5, jy=0.5)
            Ahorn.drawSprite(ctx, sprite, -3, -1, jx=0.5, jy=0.5)
            Ahorn.drawSprite(ctx, sprite, 1, 3, jx=0.5, jy=0.5)
        end
        Ahorn.Cairo.restore(ctx)
    end
end

end

File: Loenn/entities/force_dash_crystal.lua
local utils = require "utils"
local drawableSprite = require "structs.drawable_sprite"

local directions = {
    ["Right"] = "Right",
    ["Down Right"] = "Downright",
    ["Down"] = "Down",
    ["Down Left"] = "Downleft",
    ["Left"] = "Left",
    ["Up Left"] = "Upleft",
    ["Up"] = "Up",
    ["Up Right"] = "Upright",
    ["None"] = "None"
}

local forceDashCrystal = {}

forceDashCrystal.name = "vitellary/forcedashcrystal"
forceDashCrystal.depth = -100

forceDashCrystal.fieldInformation = {
    respawnTime = {
        minimumValue = 0.0
    },
    direction = {
        editable = false,
        options = directions
    }
}

forceDashCrystal.placements = {}
for _, dir in pairs(directions) do
    local placement = {
        name = string.lower(dir),
        data = {
            oneUse = false,
            direction = dir,
            respawnTime = 2.5,
            needDash = false
        }
    }

    table.insert(forceDashCrystal.placements, placement)
end

local crystalSpriteData = {
    Right = {
        ortho = true,
        rotation = 0.5 * math.pi
    },
    Downright = {
        ortho = false,
        rotation = 0.5 * math.pi
    },
    Down = {
        ortho = true,
        rotation = math.pi
    },
    Downleft = {
        ortho = false,
        rotation = math.pi
    },
    Left = {
        ortho = true,
        rotation = -0.5 * math.pi
    },
    Upleft = {
        ortho = false,
        rotation = -0.5 * math.pi
    },
    Up = {
        ortho = true,
        rotation = 0
    },
    Upright = {
        ortho = false,
        rotation = 0
    },
    None = {
        ortho = false,
        rotation = 0
    }
}

local crystalOffsets = {
    ortho = {
        {x = 0, y = -2},
        {x = -4, y = 2},
        {x = 4, y = 2}
    },
    diag = {
        {x = 2, y = -2},
        {x = -3, y = -1},
        {x = 1, y = 3}
    },
    none = {
        {x = 2, y = -2, r = 0},
        {x = 2, y = -2, r = 0.5 * math.pi},
        {x = 2, y = -2, r = math.pi},
        {x = 2, y = -2, r = -0.5 * math.pi}
    }
}

function forceDashCrystal.sprite(room, entity)
    local direction = entity.direction or "Right"
    local noDirection = direction == "None"

    local data = crystalSpriteData[direction]

    local appearance = entity.needDash and "needdash" or "dashless"
    local orientation = data.ortho and "ortho" or "diag"
    local texture = string.format("objects/crystals/forcedash/%s/%s/idle00", appearance, orientation)

    local rotation = data.rotation or 0

    local sprites = {}

    local offsets = noDirection and crystalOffsets.none or data.ortho and crystalOffsets.ortho or crystalOffsets.diag
    for _, o in ipairs(offsets) do
        local sprite = drawableSprite.fromTexture(texture, entity)
        sprite.rotation = o.r or rotation
        local cos, sin = math.cos(sprite.rotation), math.sin(sprite.rotation)

        local ox = o.x * cos - o.y * sin
        local oy = o.y * cos + o.x * sin
        sprite:addPosition(ox, oy)

        table.insert(sprites, sprite)
    end

    return sprites
end

function forceDashCrystal.selection(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    return utils.rectangle(x - 8, y - 8, 16, 16)
end

return forceDashCrystal
